apps/src/redux.js

```diff
@@ -1,21 +1,124 @@
+import reduxThunk from 'redux-thunk';
+import { connect } from 'react-redux';
 
-var redux = require('redux');
-var reduxThunk = require('redux-thunk').default;
 if (process.env.NODE_ENV !== "production") {
   var createLogger = require('redux-logger');
 }
 
+
+let reduxStore;
+let globalReducers = {};
+
+let __oldReduxStore;
+let __oldGlobalReducers;
+
+export function stubRedux() {
+  if (__oldReduxStore) {
+    throw new Error("Reduce store has already been stubbed. Did you forget to call restore?");
+  }
+  __oldReduxStore = reduxStore;
+  __oldGlobalReducers = globalReducers;
+  reduxStore = null;
+  globalReducers = {};
+}
+
+export function restoreRedux() {
+  reduxStore = __oldReduxStore;
+  globalReducers = __oldGlobalReducers;
+  __oldReduxStore = null;
+  __oldGlobalReducers = null;
+}
+
+
+/**
+ * Get a reference to our redux store. If it doesn't exist yet, create it.
+ */
+export function getStore() {
+  if (!reduxStore) {
+    reduxStore = createStore(redux.combineReducers(globalReducers));
+  }
+
+  return reduxStore;
+}
+
+/**
+ * Register a top-level reducer with the global store and get back
+ * a selector function to access the state for that reducer.
+ * @param {string} key - a unique key to identify this reducer.
+ *     The key will be used in the state object.
+ * @param {function} reducer - the reducer function
+ * @returns {function} a selector function to access the state for that reducer.
+ */
+export function registerReducer(key, reducer) {
+  return registerReducers({[key]: reducer})[key];
+}
+
+/**
+ * Register multiple top-level reducers with the global store and get back
+ * selector functions to access the state for each reducer.
+ * @param {object} reducers - an object mapping unique keys to reducer functions
+ *     The keys will be used in the state object.
+ * @returns {object} - an object mapping the keys to selector functions for accessing
+ *     the state of the corresponding reducer.
+ */
+export function registerReducers(reducers) {
+  for (var key in reducers) {
+    if (hasReducer(key)) {
+      throw new Error(`reducer with key "${key}" already registered!`);
+    }
+  }
+  Object.assign(globalReducers, reducers);
+  if (reduxStore) {
+    reduxStore.replaceReducer(redux.combineReducers(globalReducers));
+  }
+  const selectors = {};
+  for (key in reducers) {
+    selectors[key] = getReducerStateSelector(key);
+  }
+  return selectors;
+}
+
+/**
+ * @returns {function} a selector function to access the state for that reducer.
+ */
+export function getReducerStateSelector(key) {
+  return state => state[key];
+}
+
+/**
+ * @returns {function} a new connect function where the state is the state for the
+ * specified global reducer key.
+ */
+export function getConnectFunction(key) {
+  const selector = getReducerStateSelector(key);
+  return function (mapStateToProps, ...args) {
+    return connect(
+      (state, ...rest) => mapStateToProps(selector(state), ...rest),
+      ...args
+    );
+  };
+}
```

Comment:
This is an experiment that I'd love to get some feedback on. Initially I thought it was super important that various subfeatures (i.e. ducks in the redux parlance) not know about the shape of the global state object lest they end up being inherently coupled. In that scenario, it's nice to have a `connect()` function be dynamically created at configuration time when the store is first setup. Then each connected component just uses the connect function for it's corresponding subfeature.

On the other hand, having that extra layer of indirection adds some complexity, and doesn't really buy us anything right now (though it might in the future).  I think I'll probably just get rid of this function for the time being.  What do you think? How important is it that each module that uses redux be completely decoupled from knowledge of the global store?